knimin/lib/data_access.py

```diff
@@ -2588,110 +2615,168 @@ def _sample_exists(self, sample_id):
+            If at least on study ID does not exist
+            If one or more sample barcodes do not exist
         """
-        sql = """SELECT EXISTS (SELECT 1 FROM barcodes.barcode
-                                WHERE barcode = %s)"""
-        if not self._con.execute_fetchone(sql, [barcode])[0]:
-            raise ValueError('Barcode %s does not exist.' % barcode)
+        with TRN:
+            study_ids = set().union(*[x['study_ids'] for x in samples
+                                      if 'study_ids' in x])
+            for study_id in study_ids:
+                self._study_exists(study_id)
+            barcodes = [x['barcode'] for x in samples if 'barcode' in x]
+            if barcodes:
+                sql = """SELECT barcode
+                         FROM barcodes.barcode
+                         WHERE barcode IN %s"""
+                TRN.add(sql, [tuple(barcodes)])
+                res = set(barcodes) - set([x[0] for x in
+                                          TRN.execute_fetchindex()])
+                if res:
+                    raise ValueError('Barcode(s) %s do not exist.' %
+                                     ', '.join(sorted(res)))
 
     def create_samples(self, samples):
-        """Creates samples
+        """Creates samples and associate them with studies
 
         Parameters
         ----------
         samples : list of dict
-            A list of samples to create and their properties to set
-            {
-             id : str,
+            A list of samples specified by properties and study IDs
+            {id : str,
                 Assigns an ID to the sample
              is_blank : bool, optional, default: False,
                 Specifies whether the sample is blank
              barcode : str, optional,
                 Assigns a barcode to the sample
              notes : str, optional
                 Makes notes of the sample
-            }
+             study_ids : list of int
+                Associates the sample with one or more studies}
 
         Raises
         ------
         ValueError
-            If at least one sample ID already exists
+            If one or more sample IDs already exist
+            If one or more study IDs do not exist
         """
         with TRN:
+            sample_ids = [x['id'] for x in samples]
+            self._samples_exist(sample_ids, exist=False)
+            self._validate_samples(samples)
             for sample in samples:
-                sample_id = sample['id']
-                if self._sample_exists(sample_id):
-                    raise ValueError('Sample ID %s already exists.'
-                                     % sample_id)
-                barcode = sample.get('barcode')
-                if barcode:
-                    self._barcode_exists(barcode)
-                is_blank = sample.get('is_blank', False)
-                notes = sample.get('notes')
                 sql = """INSERT INTO pm.sample (sample_id, is_blank, barcode,
```

Comment:
Minor, but you can specify this SQL string outside the for loop, so you don't have to create it every time.